File: reef-wake/wake/src/main/java/org/apache/reef/wake/impl/TimerStage.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.wake.impl;

import org.apache.reef.tang.annotations.Parameter;
import org.apache.reef.wake.EventHandler;
import org.apache.reef.wake.Stage;
import org.apache.reef.wake.StageConfiguration.StageHandler;
import org.apache.reef.wake.StageConfiguration.StageName;
import org.apache.reef.wake.StageConfiguration.TimerInitialDelay;
import org.apache.reef.wake.StageConfiguration.TimerPeriod;
import org.apache.reef.wake.WakeParameters;

import javax.inject.Inject;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Stage that triggers an event handler periodically
 */
public final class TimerStage implements Stage {
  private static final Logger LOG = Logger.getLogger(TimerStage.class.getName());

  private final AtomicBoolean closed = new AtomicBoolean(false);
  private final ScheduledExecutorService executor;
  private final PeriodicEvent event = new PeriodicEvent();
  private final long shutdownTimeout = WakeParameters.EXECUTOR_SHUTDOWN_TIMEOUT;

  /**
   * Constructs a timer stage with no initial delay
   *
   * @param handler an event handler
   * @param period  a period in milli-seconds
   */
  @Inject
  public TimerStage(@Parameter(StageHandler.class) final EventHandler<PeriodicEvent> handler,
                    @Parameter(TimerPeriod.class) final long period) {
    this(handler, 0, period);
  }

  /**
   * Constructs a timer stage with no initial delay
   *
   * @param handler an event handler
   * @param period  a period in milli-seconds
   * @name name the stage name
   */
  @Inject
  public TimerStage(@Parameter(StageName.class) final String name,
                    @Parameter(StageHandler.class) final EventHandler<PeriodicEvent> handler,
                    @Parameter(TimerPeriod.class) final long period) {
    this(name, handler, 0, period);
  }

  /**
   * Constructs a timer stage
   *
   * @param handler      an event handler
   * @param initialDelay an initial delay
   * @param period       a period in milli-seconds
   */
  @Inject
  public TimerStage(@Parameter(StageHandler.class) final EventHandler<PeriodicEvent> handler,
                    @Parameter(TimerInitialDelay.class) final long initialDelay,
                    @Parameter(TimerPeriod.class) final long period) {
    this(handler.getClass().getName(), handler, initialDelay, period);
  }

  /**
   * Constructs a timer stage
   *
   * @param name         the stage name
   * @param handler      an event handler
   * @param initialDelay an initial delay
   * @param period       a period in milli-seconds
   */
  @Inject
  public TimerStage(@Parameter(StageName.class) final String name,
                    @Parameter(StageHandler.class) final EventHandler<PeriodicEvent> handler,
                    @Parameter(TimerInitialDelay.class) final long initialDelay,
                    @Parameter(TimerPeriod.class) final long period) {
    this.executor = Executors.newScheduledThreadPool(1, new DefaultThreadFactory(name));
    executor.scheduleAtFixedRate(new Runnable() {

      @Override
      public void run() {
        if (LOG.isLoggable(Level.FINEST)) LOG.log(Level.FINEST, "{0} {1}", new Object[]{name, event});
        handler.onNext(event);
      }

    }, initialDelay, period, TimeUnit.MILLISECONDS);
    StageManager.instance().register(this);
  }


  /**
   * Closes resources
   *
   * @throws Exception
   */
  @Override
  public void close() throws Exception {
    if (closed.compareAndSet(false, true)) {
      executor.shutdown();
      if (!executor.awaitTermination(shutdownTimeout, TimeUnit.MILLISECONDS)) {
        LOG.log(Level.WARNING, "Executor did not terminate in " + shutdownTimeout + "ms.");
        List<Runnable> droppedRunnables = executor.shutdownNow();
        LOG.log(Level.WARNING, "Executor dropped " + droppedRunnables.size() + " tasks.");
      }
    }
  }

}

File: lang/cs/Org.Apache.REEF.Wake/Impl/TimerStage.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.Threading;

namespace Org.Apache.REEF.Wake.Impl
{
    /// <summary>Stage that triggers an event handler periodically</summary>
    public sealed class TimerStage : IStage
    {
        // private readonly ScheduledExecutorService executor;
        private readonly Timer _timer;
        private readonly PeriodicEvent _value = new PeriodicEvent();
        private readonly IEventHandler<PeriodicEvent> _handler;

        /// <summary>Constructs a timer stage with no initial delay</summary>
        /// <param name="handler">an event handler</param>
        /// <param name="period">a period in milli-seconds</param>
        public TimerStage(IEventHandler<PeriodicEvent> handler, long period) : this(handler, 0, period)
        {
        }

        /// <summary>Constructs a timer stage</summary>
        /// <param name="handler">an event handler</param>
        /// <param name="initialDelay">an initial delay</param>d
        /// <param name="period">a period in milli-seconds</param>
        public TimerStage(IEventHandler<PeriodicEvent> handler, long initialDelay, long period)
        {
            // Core .NET only supports 32 bit timers.
            validate("initialDelay", initialDelay);
            validate("initialDelay", period);

            _handler = handler;
            _timer = new System.Threading.Timer(
                (object state) => { OnTimedEvent(_handler, _value); }, this, (int)initialDelay, (int)period);
        }

        /// <summary>
        /// Closes resources
        /// </summary>
        public void Dispose()
        {
            _timer.Dispose();
        }

        private static void OnTimedEvent(IEventHandler<PeriodicEvent> handler, PeriodicEvent value)
        {
            handler.OnNext(value);
        }

        /// <summary>
        /// Validates the input is less than Int32.MaxInt. 
        /// </summary>
        /// <param name="name">Parameter name</param>
        /// <param name="value">Parameter value</param>
        /// <exception cref="ArgumentException">Input value excess Int32.Max</exception>
        private void validate(string name, long value)
        {
            if (value > int.MaxValue)
            {
                throw new ArgumentException(string.Format(
                    "Parameter: " + name + " {0} is larger supported value {0}", int.MaxValue));
            }
        }
    }
}